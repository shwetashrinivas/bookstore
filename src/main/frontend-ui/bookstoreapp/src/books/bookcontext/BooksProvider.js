import React, { createContext, useState, useEffect } from "react";
import axios from "axios";

export const BooksContext = createContext();

const BooksProvider = ({ children }) => {
  const [books, setBooks] = useState({
    books: [],
  });

  useEffect(() => {
    const getBooks = async () => await axios.get("http://localhost:8090/books");
    getBooks()
      .then(({ data }) => {
        setBooks([...data]);
      })
      .catch(() => {});
  }, []);

  return (
    <BooksContext.Provider
      value={{
        setBooks,
        books,
      }}
    >
      {children}
    </BooksContext.Provider>
  );
};

export default BooksProvider;
