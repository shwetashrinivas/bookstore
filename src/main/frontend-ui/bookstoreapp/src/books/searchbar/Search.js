import React, { useState, useEffect, useContext } from "react";
import axios from "axios";
import "./search.css";
import { BooksContext } from "../bookcontext/BooksProvider.js";

export default function Search() {
  const { setBooks } = useContext(BooksContext);
  const [searchValue, setSearchValue] = useState("");
  useEffect(() => {
    const getSereachedBooks = setTimeout(() => {
      axios
        .get("http://localhost:8090/books/search", {
          params: {
            searchTerm: searchValue,
          },
        })
        .then((response) => {
          const { data } = response;
          console.log("response", data);
          setBooks([...data]);
        })
        .catch(() => {});
    }, 500);

    return () => clearTimeout(getSereachedBooks);
  }, [searchValue, setBooks]);

  return (
    <input
      placeholder="Search book"
      data-testid="search-bar"
      className="input-style"
      onChange={(event) => {
        setSearchValue(event.target.value);
      }}
    />
  );
}
