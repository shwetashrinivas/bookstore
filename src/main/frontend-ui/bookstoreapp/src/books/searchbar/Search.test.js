import { render, screen, fireEvent } from "@testing-library/react";
import Search from "./search.js";
import MockAdapter from "axios-mock-adapter";
import axios from "axios";
import { BooksContext } from "../bookcontext/BooksProvider.js";

const mockBookList = [
  {
    id: 1,
    title: "Tale of two cities",
    author: "Charles Dickens",
    price: 350.0,
    description:
      "The novel tells the story of the French Doctor Manette, his 18-year-long imprisonment in the Bastille in Paris, and his release to live in London with his daughter Lucie whom he had never met. The story is set against the conditions that led up to the French Revolution and the Reign of Terror.",
    coverImageUrl: "1586_f1601742134.png",
    rating: 5.0,
    quantity: 3,
  },
  {
    id: 2,
    title: "Little women",
    author: "Louisa May Alcott",
    price: 450.0,
    description:
      "It is set during and after the Civil War and tells the story of the March family, principally the four girls: Meg, Jo, Beth, and Amy. The girls struggle and grow as a family while their father is serving in the war and learn the value of hard work, self-sacrifice, and love.",
    coverImageUrl: "1586_f1601742134.png",
    rating: 5.0,
    quantity: 4,
  },
];
const mockAxios = new MockAdapter(axios);

const renderBooks = () => {
  return (
    <BooksContext.Provider
      value={{
        books: mockBookList,
        setBooks: () => {},
      }}
    >
      <Search />
    </BooksContext.Provider>
  );
};

describe("search bar", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.useFakeTimers("legacy");
    mockAxios
      .onGet("http://localhost:8090/books/search")
      .reply(200, mockBookList);
  });
  it("should rendering search bar", () => {
    const { container } = render(renderBooks());
    jest.runAllTimers();
    expect(container).toMatchSnapshot();
  });

  it("should present input search box", () => {
    render(renderBooks());
    jest.runAllTimers();
    expect(screen.getByTestId("search-bar")).toBeInTheDocument();
  });

  it("enter characters to input box", () => {
    render(renderBooks());
    jest.runAllTimers();
    fireEvent.change(screen.getByTestId("search-bar"), {
      target: { value: "abc" },
    });
    expect(screen.getByTestId("search-bar")).toHaveValue("abc");
  });

  it("when /search end point fails", () => {
    mockAxios.onGet("http://localhost:8090/books/search").reply(500, {});
    render(renderBooks());
    jest.runAllTimers();
    fireEvent.change(screen.getByTestId("search-bar"), {
      target: { value: "abc" },
    });
    expect(screen.getByTestId("search-bar")).toHaveValue("abc");
  });
});
