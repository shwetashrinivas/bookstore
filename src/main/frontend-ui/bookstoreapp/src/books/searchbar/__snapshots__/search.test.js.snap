// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search bar should rendering search bar 1`] = `
<div>
  <input
    class="input-style"
    data-testid="search-bar"
    placeholder="Search book"
  />
</div>
`;
